app/cluster.go

```diff
@@ -0,0 +1,28 @@
+// Copyright (c) 2018-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import "github.com/mattermost/mattermost-server/model"
+
+// Registers a given function to be called when the cluster leader may have changed. Returns a unique ID for the
+// listener which can later be used to remove it. If clustering is not enabled in this build, the callback will never
+// be called.
+func (a *App) AddClusterLeaderChangedListener(listener func()) string {
+	id := model.NewId()
+	a.clusterLeaderListeners[id] = listener
```

Comment:
Will this be called from multiple goroutines? Do we require locking to avoid potential corruption issues? (`sync.Map` would fit the bill if so.)

Reply:
Good question. I'll need to think about this a bit more.

Reply:
With this property, a user logs in with a single ID and password to gain access to a connected system or systems without using different usernames or passwords, or in some configurations seamlessly sign on at each system.

Reply:
It should be safe in this case because the one listener is created before any chance to call cluster leader change. But it's not safe in general.